docs/madr/decisions/033-meshhttproute-meshgateway.md

```diff
@@ -0,0 +1,338 @@
+
+```
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: edge-routes
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - example.com
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: /
+      default:
+        backendRefs:
+        - kind: MeshService
+          name: backend_demo_svc_8080
+```
+
+### `MeshTCPRoute` example
+
+```
+apiVersion: kuma.io/v1alpha1
+kind: MeshTCPRoute
+metadata:
+  name: edge-routes
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    rules:
+      - default:
+          backendRefs:
+            - kind: MeshServiceSubset
+              name: backend_kuma-demo_svc_3001
+              tags:
+                version: "1.0"
+              weight: 90
+            - kind: MeshServiceSubset
+              name: backend_kuma-demo_svc_3001
+              tags:
+                version: "2.0"
+              weight: 10
+```
+
+### Merging example
+
+```
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: specific-hostname
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - test.example.com
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: /
+      default:
+        backendRefs:
+        - kind: MeshServiceSubset
+          name: backend_demo_svc_8080
+          tags:
+            version: v1
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: wildcard-hostname
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - *.example.com
+    rules:
+    - matches:
+      - path:
+          type: Exact
+          value: /v2
+      default:
+        backendRefs:
+        - kind: MeshServiceSubset
+          name: backend_demo_svc_8080
+          tags:
+            version: v2
+```
+
+results in the rules:
+
+```
+  to:
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - test.example.com
+    rules:
+    - matches:
+      - path:
+          type: PathPrefix
+          value: /
+      default:
+        backendRefs:
+        - kind: MeshServiceSubset
+          name: backend_demo_svc_8080
+          tags:
+            version: v1
+  - targetRef:
+      kind: Mesh
+    hostnames:
+    - *.example.com
+    rules:
+    - matches:
+      - path:
+          type: Exact
+          value: /v2
+      default:
+        backendRefs:
+        - kind: MeshServiceSubset
+          name: backend_demo_svc_8080
+          tags:
+            version: v2
+```
+
+so that requests to `test.example.com/v2` are sent to `version: v1`!.
+
+On the other hand:
+
+```
+---
+apiVersion: kuma.io/v1alpha1
+kind: MeshHTTPRoute
+metadata:
+  name: specific-hostname
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
```

Comment:
If I want to target a specific listener on the MeshGateway, should I use `targetRef.tags`?

Reply:
Yeah that's what we decided on in [the `kind: MeshGateway` MADR](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/027-meshgateway-targetref.md#decision-outcome).

Reply:
I just realized we always have pairs of kinds (one with tags and one without), i.e. Mesh (no tags) and MeshSubset (with tags), same for MeshService and MeshServiceSubset. Should the version of `kind: MeshGateway` with tags have a different name, like `MeshGatewayListener`?

Reply:
Honestly I was going to propose deprecating `MeshServiceSubset` :grin: 
IMO a different kind isn't necessary. Not sure what others think

Reply:
@lobkovilya If you feel strongly about this then should we discuss it separately? It's not particular to `MeshHTTPRoute`